### ISSUE TITLE
Is the hash function slow enough?

### ISSUE THREAD

**Issue by WelliSolutions** [magento/magento2#2106]
The code for creating a hash uses a salt, but as far as I can see, it applies the SHA256 algorithm only once. Is that slow enough or should this be repeated a few times?

    public function getHash($password, $salt = false)
    {
        if ($salt === false) {
            return $this->hash($password);
        }
        if ($salt === true) {
            $salt = self::DEFAULT_SALT_LENGTH;
        }
        if (is_integer($salt)) {
            $salt = $this->randomGenerator->getRandomString($salt);
        }
        return $this->hash($salt . $password) . ':' . $salt;
    }
    public function hash($data, $version = self::HASH_VERSION_LATEST)
    {
        if (self::HASH_VERSION_MD5 === $version) {
            return md5($data);
        }
        return hash('sha256', $data);
    }

**Comment by orlangur** [comment]
It's a big surprise to me that having PHP 5.5 as a minimum required version Magento 2 does not use the [new hashing API](https://wiki.php.net/rfc/password_hash).

After looking into http://www.php.net/manual/en/function.password-hash.php:
> Warning
The salt option has been deprecated as of PHP 7.0.0. It is now preferred to simply use the salt that is generated by default.

I believe contemporary answer to your question is "no salt needed at all, simply use latest hashing API".

**Comment by lepe** [comment]
Agree. Perhaps using PBKDF-2 or bcrypt instead of SHA256?

**Comment by antonkril** [comment]
We're looking at it. I will let you know when we have solution.

**Comment by Flyingmana** [comment]
related to #992

**Comment by davidalger** [comment]
Closing as a duplicate of #992. There is already an internal issue open for using the native PHP Password API.
